standardization/src/main/scala/za/co/absa/enceladus/standardization/interpreter/stages/TypeParser.scala

```diff
@@ -15,185 +15,331 @@
+          ).otherwise(
+            typedLit(Seq[ErrorMessage]())
+          )
+        )
+
+        val std = when((size(err) === lit(0)) and (column isNotNull),
+          primitiveCastLogic
+        ).otherwise(
+          // scalastyle:off null
+          when(size(err) === lit(0), null
+          // scalastyle:on null
+          ).otherwise( Defaults.getDefaultValue(field) )
+        ) as field.name
+
+        ParseOutput(std, err)
+      }
+    }
+
+    trait SimplePW extends  PrimitivePW {
+      override def getPrimitiveCastLogic: Column = column.cast(field.dataType)
+    }
+
+    case class NumericPW(field: StructField, path: String, origSchema: StructType, parent: Parent) extends SimplePW
+
+    case class StringPW(field: StructField, path: String, origSchema: StructType, parent: Parent) extends SimplePW
+
+    case class BooleanPW(field: StructField, path: String, origSchema: StructType, parent: Parent) extends SimplePW
+
+    /**
+      * Timestamp conversion logic
+      * Original type | TZ in pattern/without TZ        | Has default TZ
+      * ==============|=================================|============================================================
+      * Fractional    | ->Decimal->String->to_timestamp | ->Decimal->String->to_timestamp->to_utc_timestamp
+      * Decimal       | ->String->to_timestamp          | ->String->to_timestamp->to_utc_timestamp
+      * String        | ->to_timestamp                  | ->to_timestamp->to_utc_timestamp
+      * Timestamp     | O                               | ->to_utc_timestamp
+      * Date          | ->to_timestamp(no pattern)      | ->to_utc_timestamp
+      * Other         | ->String->to_timestamp          | ->String->to_timestamp->to_utc_timestamp
+      *
+      *
+      * Date conversion logic
+      * Original type | TZ in pattern/without TZ        | Has default TZ (the last to_date is always without pattern)
+      * ==============|=================================|============================================================
+      * Float         | ->Decimal->String->to_date      | ->Decimal->String->to_timestamp->to_utc_timestamp->to_date
+      * Decimal       | ->String->to_date               | ->String->->to_timestamp->->to_utc_timestamp->to_date
+      * String        | ->to_date                       | ->to_timestamp->->to_utc_timestamp->to_date
+      * Timestamp     | ->to_date                       | ->to_utc_timestamp->to_date
+      * Date          | O                               | ->to_utc_timestamp->to_date
+      * Other         | ->String                        | ->String->to_timestamp->to_utc_timestamp->to_date
+      */
+
+    trait DateTimePW extends PrimitivePW {
+      protected def basicCastFunction: (Column, String) => Column  //for epoch casting
+      protected def epochPattern: String
+      protected val pattern: DateTimePattern = DateTimePattern(field)
+      protected lazy val defaultTimeZone: Option[String] = pattern.defaultTimeZone
```

Comment:
Why are there so many `lazy val`s everywhere? I feel like this would break in very unexpected ways, very easily.

Reply:
By making the code difficult to debug, reason about and change in any way. Spread out reliance on `lazy val`s is a bit of a code smell to me.